modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.prepare;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.fun.SqlCase;
+import org.apache.calcite.sql.fun.SqlLibraryOperators;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Ignite SQL call rewrite table. Performs unconditional rewrites for some predefined Calcite SQL operators,
+ * which can't be extended other ways by Ignite.
+ */
+public class IgniteSqlCallRewriteTable {
+    /** Instance. */
+    public static final IgniteSqlCallRewriteTable INSTANCE = new IgniteSqlCallRewriteTable();
+
+    /** Registered rewriters map. */
+    private final Map<SqlOperator, BiFunction<SqlValidator, SqlCall, SqlCall>> map = new HashMap<>();
+
+    /** */
+    private IgniteSqlCallRewriteTable() {
+        register(SqlLibraryOperators.NVL, IgniteSqlCallRewriteTable::nvlRewriter);
+        register(SqlLibraryOperators.DECODE, IgniteSqlCallRewriteTable::decodeRewriter);
+    }
+
+    /** Registers rewriter for SQL operator. */
+    public void register(SqlOperator operator, BiFunction<SqlValidator, SqlCall, SqlCall> rewriter) {
+        map.put(operator, rewriter);
```

Comment:
Suggestion: any thread-safe implementation. We can't be sure that custom rewriters are registered from the same thread.

Reply:
Fixed